lib/dry/system/config/component_dirs.rb

```diff
@@ -1,49 +1,181 @@
           super
           @dirs = source.dirs.dup
         end
 
-        def add(path_or_component_dir, &block)
-          if path_or_component_dir.is_a?(ComponentDir)
-            add_component_dir(path_or_component_dir)
-          else
-            build_and_add_component_dir(path_or_component_dir, &block)
+        # Adds and configures a component dir
+        #
+        # @param path [String] the path for the component dir, relative to the configured
+        #   container root
+        #
+        # @yieldparam dir [ComponentDir] the component dir to configure
+        #
+        # @return [ComponentDir] the added component dir
+        #
+        # @example
+        #   component_dirs.add "lib" do |dir|
+        #     dir.default_namespace = "my_app"
+        #   end
+        #
+        # @see ComponentDir
+        def add(path)
+          raise ComponentDirAlreadyAddedError, path if dirs.key?(path)
+
+          dirs[path] = ComponentDir.new(path).tap do |dir|
+            apply_defaults_to_dir(dir)
+            yield dir if block_given?
           end
         end
 
+        # Returns the added component dirs, with default settings applied
+        #
+        # @return [Hash<String, ComponentDir>] the component dirs as a hash, keyed by path
+        def dirs
+          @dirs.each { |_, dir| apply_defaults_to_dir(dir) }
+        end
+
+        # Returns the added component dirs, with default settings applied
+        #
+        # @return [Array<ComponentDir>]
         def to_a
           dirs.values
         end
 
+        # Calls the given block once for each added component dir, passing the dir as an
+        # argument.
+        #
+        # @yieldparam dir [ComponentDir] the yielded component dir
         def each(&block)
           to_a.each(&block)
         end
 
         private
 
-        def build_and_add_component_dir(path, &block)
-          add_component_dir(ComponentDir.new(path, &block))
+        # Apply default settings to a component dir. This is run every time the dirs are
+        # accessed to ensure defaults are applied regardless of when new component dirs
+        # are added. This method must be idempotent.
+        #
+        # @return [void]
+        def apply_defaults_to_dir(dir)
+          dir.config.values.each do |key, value|
+            if configured?(key) && !dir.configured?(key)
+              dir.public_send(:"#{key}=", public_send(key))
+            end
+          end
         end
 
-        def add_component_dir(dir)
-          raise ComponentDirAlreadyAddedError, dir.path if dirs.key?(dir.path)
+        # Returns true if a setting has been explicitly configured, and not returning just
+        # a default value.
+        #
+        # This is used to determine which settings should be applied to added component
+        # dirs as additional defaults.
+        #
+        # @api private
+        def configured?(key)
+          config._settings[key].input_defined?
+        end
```

Comment:
I had to dig into some dry-system internals here, but it _vastly_ simplifies the logic of `apply_defaults_to_dir` above.

Previously I was having to inspect the current values for both the system-wide and individual dir settings, the compare them to the settings own intrinsic default value, and then try and work out from those whether or not to apply the system-wide default. But this would fall down in the case of having a system-wide default that you want to _opt out of_ by explicitly configuring your individual dir to use the default value again.

For example:

```ruby
component_dirs.default_namespace "my_app"

component_dirs.add do |dir|
  # Explicitly provide nil here, even though it is already the "natural" default for this
  # setting, to opt out of the system-wide default configured above
  dir.default_namespace = nil
end
```

IMO, this is a reasonable behavior to support, and without the `#configured?` methods I added to `ComponentDirs` and `ComponentDir`, it would not be possible to achieve this.

And really, the best case to make for this approach is that `#apply_defaults_to_dir` is just so easy to read! `if configured?(key) && !dir.configured?(key)` – simple!

Reply:
> I had to dig into some dry-system internals here

@timriley you mean dry-configurable internals, right?

Reply:
@solnic Oops, yes, you're right, I did mean dry-configurable internals :)